### ISSUE TITLE
Investigating font atlas texture mismatching runtime data (continued)

### ISSUE THREAD

**Issue by rokups** [ocornut/imgui#2245]
... continuing from https://github.com/ocornut/imgui/issues/1542#issuecomment-448134844

Thanks for taking further interest in the issue @ocornut :)

To recap: my custom imgui renderer implementation for Urho3D had a problem, where resizing window was fine, but restoring a custom window size on startup corrupted font texture.

![](https://user-images.githubusercontent.com/19151258/50015304-e7141900-ffce-11e8-9099-e0a3caa0f7ea.png)

This is how i handled screen resizing:
1. [Subscribe to E_SCREENMODE event](https://github.com/rokups/Urho3D/blob/8f885913004bc4bd387d1805dde7bae3ae7aa1b6/Source/Urho3D/SystemUI/SystemUI.cpp#L91) which is sent every time window is resized.
2. [Event handler](https://github.com/rokups/Urho3D/blob/8f885913004bc4bd387d1805dde7bae3ae7aa1b6/Source/Urho3D/SystemUI/SystemUI.cpp#L113-L133) updates projection matrix.

[Rendering code](https://github.com/rokups/Urho3D/blob/8f885913004bc4bd387d1805dde7bae3ae7aa1b6/Source/Urho3D/SystemUI/SystemUI.cpp#L214-L313) is here.

The fix: Also call [ReallocateFontTexture()](https://github.com/rokups/Urho3D/blob/8f885913004bc4bd387d1805dde7bae3ae7aa1b6/Source/Urho3D/SystemUI/SystemUI.cpp#L340-L373) during `E_SCREENMODE` event. Font texture is re-baked and re-uploaded to GPU at this step.

I suspect GPU reupload is reason why this fixes the problem. Any idea this could be the reason? I am testing on linux/opengl.

**Comment by ocornut** [comment]
Thanks for the clarification. So - this is actually a non issue here :)
I thought you were using one of the example/ back-ends.

In this situation my guess is that:
- Either your graphics engine has this requirement were textures are invalidated when resizing the screen (or a variation on that) so this is entirely something to deal with on your engine side, as you did.
- Either for whatever reason your code was uploading to the GPU a texture that doesn't correspond to the final texture used by the font atlas (e.g. Default font / Build / Upload to GPU / User code adds new fonts / Build / Forget to upload to GPU again!).

Either way this should not be be specific to the Viewport branch and I suspect you would have had the same issue in Master.

**Comment by rokups** [comment]
I little more investigation made the whole situation clear. It indeed was caused by inner workings of the engine. Restoring window size early at runtime caused engine to drop GPU texture. What is visible in the screenshot - that is uninitialized GPU memory as far as i understand. I also confirmed that window resizing during normal runtime did not cause corruption because for some reason engine was no longer freeing font atlas texture and thus it remained valid.

Thank you for helping to unravel this mystery. Hopefully this bit will be useful for some reader form the future bumping into similar issue by accident.

tl;dr;
When resizing main viewport it is enough to update `ImGui::GetIO().DisplaySize` and recalculate projection matrix using new viewport size. Make sure your font atlas texture is reuploaded to GPU if window resize deleted it.
